merkle/smt/hstar3.go

```diff
@@ -70,32 +70,43 @@ func NewHStar3(updates []NodeUpdate, hash HashChildrenFn, depth, top uint) (HSta
 	return HStar3{upd: updates, hash: hash, depth: depth, top: top}, nil
 }
 
-// Prepare returns the set of all the node IDs that the Update method will load
-// in order to compute node hash updates from the initial tree depth up to the
-// top level specified in the constructor. It may be useful for constructing a
-// NodeAccessor, e.g. by batch-reading the nodes from elsewhere.
-//
-// Note: The returned map could have bool value type, but []byte allows the
-// caller to reuse this map for filling in the hashes for the Update method.
+// Prepare returns the list of all the node IDs that the Update method will
+// load in order to compute node hash updates from the initial tree depth up to
+// the top level specified in the constructor. It may be useful for creating a
+// a NodeAccessor, e.g. by batch-reading the nodes from elsewhere.
 //
 // TODO(pavelkalinnikov): Return only tile IDs.
-func (h HStar3) Prepare() map[tree.NodeID2][]byte {
-	ids := make(map[tree.NodeID2][]byte)
+func (h HStar3) Prepare() []tree.NodeID2 {
+	empty := tree.NodeID2{}
+	ids := []tree.NodeID2{empty} // Start with a dummy sentinel value.
+	pos := make([]int, h.depth-h.top)
```

Comment:
The relationship between pos and ids is not immediately clear. Its meaning can be derived, but I'd like to see the loop invariant (https://en.wikipedia.org/wiki/Loop_invariant) stated. Things that I had to work out:
1) All elements in `pos` are effectively pointing at the empty sentinel value in `ids` to begin
2) You fill out `pos` from back-to-front
3) There will always be at least one element in `pos` which equals `len(ids)-1` (i.e. points to the last valid index where `ids[i]` is valid)
4) The whole reason that this trick with `pos` works is precisely because the nodes are already sorted left-to-right

Reply:
Done. From your list, the items 2 is irrelevant, and 3 needed a bit of generalization. LMK how that looks to you now.